ledger/nomos-ledger/src/cryptarchia/mod.rs

```diff
@@ -188,6 +188,60 @@ impl LedgerState {
             .update_nonce(proof.entropy(), slot))
     }
 
+    pub fn try_apply_tx<Id, Constants: GasConstants>(
+        mut self,
+        tx: impl AuthenticatedMantleTx,
+    ) -> Result<(Self, Value), LedgerError<Id>> {
+        let mut balance: u64 = 0;
+        let mut pks: Vec<[u8; 32]> = vec![];
+        let ledger_tx = &tx.mantle_tx().ledger_tx;
+        for input in &ledger_tx.inputs {
+            let note;
+            (self.utxos, note) = self
+                .utxos
+                .remove(input)
+                .map_err(|_| LedgerError::InvalidNote(*input))?;
+            balance = balance
+                .checked_add(note.value)
+                .ok_or(LedgerError::Overflow)?;
+            pks.push(note.pk.into());
+        }
+
+        if !tx.ledger_tx_proof().verify(&ZkSignaturePublic {
+            pks,
+            tx_hash: tx.hash().into(),
+        }) {
+            return Err(LedgerError::InvalidProof);
+        }
+
+        for (i, output) in ledger_tx.outputs.iter().enumerate() {
+            if output.value == 0 {
+                return Err(LedgerError::ZeroValueNote);
+            }
+            balance = balance
+                .checked_sub(output.value)
+                .ok_or(LedgerError::InsufficientBalance)?;
+            self.utxos = self
+                .utxos
+                .insert(
+                    Utxo {
+                        tx_hash: tx.hash(),
+                        output_index: i,
+                        note: *output,
+                    }
+                    .id(),
+                    *output,
+                )
```

Comment:
So after we compute the note id, we throw away the index and tx hash? So we won't be able to re-construct the utxo from the information stored in the node right? we will only be able to verify a utxo by checking it against the note id?

Reply:
You can reconstruct every utxo if you have the full blockchain history (which we do), but that's not of particular interest in the ledger and we don't keep it ready in memory. Is there a usecase where do you think it's useful?